crates/terminal/src/terminal.rs

```diff
@@ -1092,7 +1104,83 @@ impl Terminal {
         self.write_bytes_to_pty(input);
     }
 
+    pub fn toggle_vi_mode(&mut self) {
+        self.vi_mode = !self.vi_mode;
+        self.events.push_back(InternalEvent::ToggleViMode);
+    }
+
+    pub fn vi_motion(&mut self, keystroke: &Keystroke) {
+        let motion: Option<ViMotion> = match keystroke.key.as_str() {
+            "h" => Some(ViMotion::Left),
+            "j" => Some(ViMotion::Down),
+            "k" => Some(ViMotion::Up),
+            "l" => Some(ViMotion::Right),
+            "w" => Some(ViMotion::WordRight),
+            "b" => Some(ViMotion::WordLeft),
+            "e" => Some(ViMotion::WordRightEnd),
+            "%" => Some(ViMotion::Bracket),
+            "$" => Some(ViMotion::Last),
+            "0" => Some(ViMotion::First),
+            "^" => Some(ViMotion::FirstOccupied),
+            _ => None,
+        };
+
+        if keystroke.key.as_str() == "v" {
+            let point = self.last_content.cursor.point;
+            let selection_type = SelectionType::Simple;
+            let side = AlacDirection::Right;
+            let selection = Selection::new(selection_type, point, side);
+            self.events
+                .push_back(InternalEvent::SetSelection(Some((selection, point))));
+            return;
+        }
+
+        if keystroke.key.as_str() == "escape" {
+            self.events.push_back(InternalEvent::SetSelection(None));
+            return;
+        }
+
+        if keystroke.key.as_str() == "y" {
+            self.events.push_back(InternalEvent::Copy);
+            self.events.push_back(InternalEvent::SetSelection(None));
+            return;
+        }
+
+        if keystroke.key.as_str() == "i" {
+            if self.vi_mode {
+                self.scroll_to_bottom();
+                self.toggle_vi_mode();
+            }
+            return;
+        }
+
+        if let Some(motion) = motion {
+            let cursor = self.last_content.cursor.point;
+            let cursor_pos = Point {
+                x: Pixels::from(cursor.column.0 as f32 * self.last_content.size.cell_width),
+                y: Pixels::from(cursor.line.0 as f32 * self.last_content.size.line_height),
+            };
+            self.events
+                .push_back(InternalEvent::UpdateSelection(cursor_pos));
+            self.events.push_back(InternalEvent::ViMotion(motion));
```

Comment:
Again: won't we end up here again?

Reply:
Not sure what you mean here to be honest, UpdateSelection is the same code called by mouse_drag to update what's currently highlighted and what you can copy. This code simply leverages that and uses vi_motion.

Reply:
Sorry, I misunderstood the original code!